### PR TITLE
Preserve non-ASCII whitespace characters in HTML

### LINKED ISSUE #7103: U+2005 etc. is (sometimes) stripped from html
**Prettier 1.19.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAeADgPlQZ3QQykwA1UB6PQzAAkFACWiAJ3utwKIE1zKjysAdKBhJ0GzWtS5kBUEABoQEdDACW0HMlD5GjCAHcACtoQaU+AG4QVAE3kgARo3xgA1nBgBlAmBVQA5sgwjACucAoAFjAAtgA2AOrhKvCUYHAeJkkq5kkAnsjgOBoKvjhwjDAGTn5R+MgAZvgxpQoAVjgAHgBCTq7uHvhRcAAyvnD1jc0gbe0evn4xcACKwRDw401hIASMpYz5kbF26Iy+MHE2MOHIABwADArHEKVxTuj5x3C75mMKAI4r8EqSlMIHwOAAtFA4HBrDC7Iw4P8VAjKvhqrUkA0NgpSlEVIEQpscHMFstVmNMRNNjB8PZztZLsgAEwKIL4FQxOYAYQgURq+Sg0B+IGCpQAKrTTFjJuZQgBJKCw2AeMAnZQAQUVHhgOQW61KAF8DUA)
```sh
# Options (if any):
--single-quote
```

**Input:**
```html
<p><span>X</span>   or   <span>Y</span></p>
<p>X   or   Y</p>
```

**Output:**
```html
<p><span>X</span> or <span>Y</span></p>
<p>X   or   Y</p>
```

**Expected behavior:**

U+2005 should not be turned into a regular space.

From a quick test, this also happens for similar characters in the [Unicode General Punctuation block](https://en.wikipedia.org/wiki/General_Punctuation).

### PR DESCRIPTION
**Pull request by fisker**
Fix 4 kinds of space detection.

*In all examples, the first one has a `U+2005`, and it format like the third one, the new one format just like the second one*

OLD:

**Prettier 2.0.5**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAeAhAWkwAgJZQDOeAJnDtgHwA6UqJeAbpaoQA4CGUlOgoAQ6oA9Oy4tBDZrXpMWI7jgD6A4Z25CJNOhtaqeOZXLEbaUrLgBGcGAHc4CCpk3TmqPGLd8Bbod-Eyp2r4eSq7uRv5aMqE+el5h6hGm2PhQZPCwDk6EYABOeGwwODAAnmxwALzUIPAAHjCCAK5QANZQENZQVTgANlwA5pUgTa3tnSCanvpT+oKCOJAAtgsIMLTT6ziz8xBLK2sbU1uLy7D7B5tzx3t0wrn5MFl3BUWlFVW19cNtHV29UANVL6jKoTRQHI47E6rKDnGaXSHXWEXba7U4w2EQ1HQoTZPIFR54wolMqDD6NFrfMY9fqDIE-cZncHwrGMjaYqGs9bsxEY5kcm64+6aJK4YgLNjdchwSVQzIBGSWABmEBycB0ol4ON0HEV8ByOr1Brg+t1xqNJsNpotZqt5rttodlqdNud1rdCUkkWYSpVasMCi1ontruDLrDbtDEcd4cjsejOQ9Tg0PtV6vkge4cZD8azMZz+ezhfDiZAABoQBACnhoIRkKAODkcu0AAoNhC1lAcbrWDjFWvl8z6sDNKwAZU4YAIfWQMByDTg5YAFjAFt0AOqLvDwERgOCj9tbphb4rIcCEfsgAiEY0wZv6voLDjIRVd6-lgBWhBqACEhyOYKOHDLAAMgQcDPq+C4gJ+NSjlOkoAIoNBA8AQd0b4gJwOTXjkp7LquZaYXksBrqQMCLsgAAcAAM5ZsE215rvqbCnvRcA4Yw4HlgAjsh8B3pWHYgBwhCYFAdhkCQhGqrxeCqneHAPk+SAvuhUHXgseAznO6nwXASEoeBKmQeWMAcOYpEkORyAAEymfqeDdFOADCkLKSA7EAKyEQ014ACrmR2qkYYw84AJKpCso5PDAACCqSjiUkpodeAC+qVAA)
```sh
--parser html
```

**Input:**
```html
<!-- inside -->
<div><span>   </span></div>
<div><span> _ </span></div>
<div><span>   </span></div>

<!-- between -->
<div><i></i>   <i></i></div>
<div><i></i> _ <i></i></div>
<div><i></i>   <i></i></div>

<!-- indetent -->
<script type="text/unknown" lang="unknown">
          // comment
          // comment
          // comment
          // comment
</script>
<script type="text/unknown" lang="unknown">
  _       // comment
          // comment
          // comment
          // comment
</script>
<script type="text/unknown" lang="unknown">
          // comment
          // comment
          // comment
          // comment
</script>

<!-- simple element -->
<div>before<span> </span>afterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafter</div>
<div>before<span>_</span>afterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafter</div>
<div>before<span> </span>afterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafter</div>
```

**Output:**
```html
<!-- inside -->
<div><span> </span></div>
<div><span> _ </span></div>
<div><span> </span></div>

<!-- between -->
<div><i></i> <i></i></div>
<div><i></i> _ <i></i></div>
<div><i></i> <i></i></div>

<!-- indetent -->
<script type="text/unknown" lang="unknown">
  // comment
  // comment
  // comment
  // comment
</script>
<script type="text/unknown" lang="unknown">
  _       // comment
          // comment
          // comment
          // comment
</script>
<script type="text/unknown" lang="unknown">
  // comment
  // comment
  // comment
  // comment
</script>

<!-- simple element -->
<div>
  before<span
  > </span
  >afterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafter
</div>
<div>
  before<span>_</span>afterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafter
</div>
<div>
  before<span
  > </span
  >afterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafterafter
</div>

```
NEW:

[Playground](https://deploy-preview-8137--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAeAhAWkwAgJZQDOeAJnDtgHwA6UqJeAbpaoQA4CGUlOgoAQ6oA9Oy4tBDZrXpMWI7jgD6A4Z25CJNOhtaqeOZXLEbaUrLgBGcGAHc4CCpk3TmqPGLd8Bbod-Eyp2r4eSq7uRv5aMqE+el5h6hGm2PhQZPCwDk6EYABOeGwwODAAnmxwALzUIPAAHjCCAK5QANZQENZQVTgANlwA5pUgTa3tnSCanvpT+oKCOJAAtgsIMLTT6ziz8xBLK2sbU1uLy7D7B5tzx3t0wrn5MFl3BUWlFVW19cNtHV29UANVL6jKoTRQHI47E6rKDnGaXSHXWEXba7U4w2EQ1HQoTZPIFR54wolMqDD6NFrfMY9fqDIE-cZncHwrGMjaYqGs9bsxEY5kcm64+6aJK4YgLNjdchwSVQzIBGSWABmEBycB0ol4ON0HEV8ByOr1Brg+t1xqNJsNpotZqt5rttodlqdNud1rdCUkkWYSpVasMCi1ontruDLrDbtDEcd4cjsejOQ9Tg0PtV6vkge4cZD8azMZz+ezhfDiZAABoQBACnhoIRkKAODkcu0AAoNhC1lAcbrWDjFWvl8z6sDNKwAZU4YAIfWQMByDTg5bgC0sJDIJAAMv0Ghw+nAAGIqhYcGAwKfIEAcBowCBlkAACxgC26AHU73h4CIwHBR+330x38U55gIQ-YgAQhDGjAzb6n0R7IIqXYQeWABWhA1AAQkOI4wKOHDLOuBBwPBiELiAqE1KOU6SgAig0EDwMR3RISAnA5BBOTng+T63mweSwM+pAwHeyAABwAAzlrxEAQc++psOevFwOxjBEeWACOdHwNBlYdhehCYFAdhrreqoaXgqrQTucFIAhTGkRBCx4DOc72VRcC0fRRE2SR5YwBw5gCSQQnIAATL5+p4N0U4AMKQhw55KQArLeDQQQAKv5Ha2cxjDzgAkqkKyjk8MAAIKpKOJSSoxEEAL61UAA)

Fix #7103 

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by brodybits** [comment]
I don't see any U+2005 characters in the test. Am I missing anything?

File: tests/html_whitespace/jsfmt.spec.js [hunk 1]
```diff
@@ -1 +1,56 @@
...
+        "        \u0005\u0005 ",
+        "        \u0005        \u0005 ",
+      ].map((textContent) => `<div>${textContent}</div>`),
+
+      ...[
+        "\u0005",
+        " \u0005        ",
+        "        \u0005\u0005 ",
+        "        \u0005        \u0005 ",
+      ].map((textContent) => `<span>${textContent}</span>`),
```
Comment (fisker) [inline tests/html_whitespace/jsfmt.spec.js:49]: @brodybits ? so many

File: tests/html_whitespace/jsfmt.spec.js [hunk 2]
```diff
@@ -1 +1,56 @@
...
+
+      // #7103
+      "<p><span>X</span> \u0005or\u0005 <span>Y</span></p><p>X \u0005or\u0005 Y</p>",
```
Comment (brodybits) [inline tests/html_whitespace/jsfmt.spec.js:52]: This change seems to reproduce #7103 for me:

```suggestion
      "<p><span>X</span> \u0005or\u0005 <span>Y</span></p><p>X \u2005or\u2005 Y</p>",
```

Comment (fisker) [reply tests/html_whitespace/jsfmt.spec.js:52]: Sorry. typo

Comment (fisker) [reply tests/html_whitespace/jsfmt.spec.js]: Right, seem the `U+2005` is lost, I'll look into this. Thank you!

File: changelog_unreleased/html/pr-8137.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+// Prettier master
+require("prettier")
+  .format("<i /> \u2005 <i />", { parser: "html" })
+  .slice(5, -6)
+  .split("")
+  .map((c) => c.charCodeAt(0).toString(16));
+
+// -> [ '20', '2005', '20' ]
+```
```
Comment (fisker) [inline changelog_unreleased/html/pr-8137.md:24]: Is it okay to use js code show the difference?

**Comment by fisker** [comment]
`"<span> \u2005 </span>"` still not right.

**Comment by pkra** [comment]
@fisker thanks for digging into this -- much appreciated!

File: src/language-html/preprocess.js [hunk 4]
```diff
@@ -176,8 +176,8 @@ function mergeSimpleElementIntoText(ast /*, options */) {
     node.attrs.length === 0 &&
     node.children.length === 1 &&
     node.firstChild.type === "text" &&
-    // \xA0: non-breaking whitespace
-    !/[^\S\xA0]/.test(node.children[0].value) &&
+    // https://infra.spec.whatwg.org/#ascii-whitespace
+    !/[^\t\n\f\r ]/.test(node.children[0].value) &&
```
Comment (fisker) [inline src/language-html/preprocess.js:180]: I'm not sure about this part, will find something to test tomorrow, it's late in my location.

File: changelog_unreleased/html/pr-8137.md (hunk 3 above)
Comment (thorn0) [reply changelog_unreleased/html/pr-8137.md:24]: Totally okay.

File: src/language-html/preprocess.js [hunk 5]
```diff
@@ -313,7 +317,7 @@ function extractWhitespaces(ast /*, options*/) {
       node.children.length === 0 ||
       (node.children.length === 1 &&
         node.children[0].type === "text" &&
-        node.children[0].value.trim().length === 0)
+        htmlTrim(node.children[0].value).length === 0)
```
Comment (evilebottnawi) [inline src/language-html/preprocess.js:320]: Good job, `.trim()` is not safe in some cases, maybe we should look at `.trim()` in all code, we can have other problems, for example for `css`

**Comment by fisker** [comment]
I'm going to use #8142 to assert output, snapshots hard to read.

**Comment by fisker** [comment]
Wow, I read the source again, there are more cases....

**Comment by fisker** [comment]
I've fix all whitespace detection, changes see the PR description.

File: src/language-html/utils.js [hunk 6]
```diff
@@ -14,6 +14,28 @@ const htmlElementAttributes = require("html-element-attributes");
 const HTML_TAGS = arrayToMap(htmlTagNames);
 const HTML_ELEMENT_ATTRIBUTES = mapObject(htmlElementAttributes, arrayToMap);
 
+// https://infra.spec.whatwg.org/#ascii-whitespace
+const HTML_SPACES = new Set(["\t", "\n", "\f", "\r", " "]);
```
Comment (thorn0) [inline src/language-html/utils.js:18]: The spec calls it "whitespace" (always singular, like "code"), not "spaces". So let's call it "whitespace" too. `HTML_WHITESPACE`, `trimLeadingHtmlWhitespace` (`htmlTrimStart` looked good to me too, BTW), etc.

The word "spaces", at least for me, means only this character: " " (e.g. in this sentence: _Whitespace may consist of spaces, tab characters, and line separators._).

And it's "trailing", not "tailing".

Comment (fisker) [reply src/language-html/utils.js:18]: I'll fix

File: src/language-html/utils.js [hunk 7]
```diff
@@ -14,6 +14,27 @@ const htmlElementAttributes = require("html-element-attributes");
 const HTML_TAGS = arrayToMap(htmlTagNames);
 const HTML_ELEMENT_ATTRIBUTES = mapObject(htmlElementAttributes, arrayToMap);
 
+// https://infra.spec.whatwg.org/#ascii-whitespace
+const HTML_WHITESPACE = new Set(["\t", "\n", "\f", "\r", " "]);
+const htmlTrimStart = (string) => string.replace(/^[\t\n\f\r ]+/, "");
+const htmlTrimEnd = (string) => string.replace(/[\t\n\f\r ]+$/, "");
+const htmlTrim = (string) => htmlTrimStart(htmlTrimEnd(string));
+const htmlTrimLineByLine = (string) =>
+  string.replace(/^[\t\f\r ]*?\n|\n[\t\f\r ]*?$/g, "");
```
Comment (fisker) [inline src/language-html/utils.js:23]: @thorn0 

This function seem not "trimLineByLine",

old code `value.replace(/^\s*?\n|\n\s*?$/g, "")`,

more like `removeEmptyLine`?

Comment (fisker) [reply src/language-html/utils.js:23]: No it's not, I really don't understand...

Comment (fisker) [reply src/language-html/utils.js:23]:
```
`

\u0020

  a \u0020
     
b \u0020
\u0020
c
\u0020
`.replace(/^\s*?\n|\n\s*?$/g, "").replace(/ /g, '_')
"
_

__a__
_____
b__
___
c"
```

Comment (fisker) [reply src/language-html/utils.js:23]: `\u0020\u0020\u0020\u0020

\u0020\u0020

  a \u0020
     
b \u0020
\u0020
c
\u0020\u0020\u0020
`.replace(/^\s*?\n|\n\s*?$/g, "").replace(/ /g, '_')
"
__

__a__
_____
b__
___
c"

`trimFirstLineAndTrailingWhitespace`?

Comment (fisker) [reply src/language-html/utils.js]: Updated to `htmlTrimFirstLine` and `htmlTrimEnd` seems work https://github.com/prettier/prettier/pull/8137/commits/9a9d36f3ed9e5cb994e09146efcad3704943efc5

Comment (thorn0) [reply src/language-html/utils.js]: Doesn't look like an equivalent replacement.

Comment (fisker) [reply src/language-html/utils.js]: Let  me  ask the  author :)

@ikatyang , you first wrote this  in

https://github.com/prettier/prettier/commit/b52be6bdcad89c1af1239cf10aed4d235a5c4c20#diff-3e8cd8ca0d07925abf364c9d8a413220R143


Do  you  still  remember what is suppose to do ?

Comment (thorn0) [reply src/language-html/utils.js]: Okay, figured it out. This code removes leading and trailing blank lines (lines consisting only of whitespace), but preserves the leading whitespace of the first non-blank line and the trailing whitespace of the last non-blank line. The leading whitespace is needed to correctly determine the indentation of scripts (related to #8147, that's why vbscript is indented). And the trailing whitespace doesn't really seem to be of any use, so your replacement looks good. I'd only suggest renaming `htmlTrimFirstLine` to `htmlTrimLeadingBlankLines`.

I'm still not sure it's okay to indent those scripts, but so far nobody complained, and that's a separate issue anyway.

**Comment by thorn0** [comment]
I found one more place that can be improved WRT to whitespace:

**`printer-html.js`**
```js
                      new RegExp(
                        `\\n\\s{${
                          options.tabWidth *
                          countParents(
                            path,
                            (n) => n.parent && n.parent.type !== "root"
                          )
                        }}$`
                      ).test(node.lastChild.value)
```
`\\n\\s` should be changed to `\\n[\\t ]`.

**Comment by fisker** [comment]
> I found one more place that can be improved WRT to whitespace:
> 
> **`printer-html.js`**
> 
> ```js
>                       new RegExp(
>                         `\\n\\s{${
>                           options.tabWidth *
>                           countParents(
>                             path,
>                             (n) => n.parent && n.parent.type !== "root"
>                           )
>                         }}$`
>                       ).test(node.lastChild.value)
> ```
> 
> `\\n\\s` should be changed to `\\n[\\t ]`.

This one seems hard to test, I just changed it. 
It also looks not right to me, `useTabs` might need consider.
It's hard to reach anyway.

**Comment by thorn0** [comment]
That regexp indeed can be improved (to not count tabs the same way as spaces), but at least now it won't match non-ASCII whitespace.

**Comment by pkra** [comment]
Thanks again to everyone involved for making this happen!

**Pull request merged**